### DISCUSSION TITLE
Squash instructions in Dockerfile

### DISCUSSION THREAD

**Discussion by jackcaron** [docker/extension-ideas#130] [Ideas]
Right now, the `squash` feature is experimental and also squash all layers into a single one.

My suggestion is to add instructions in a `Dockerfile` on what to squash.

Ex.: if I have a set of instructions like this:
```
RUN wget -qO- "https://mystorage.com/some_debs_files.tar.xz" | tar xJf - -C . \
  && apt-get -qq update && apt-get -qq install -y \
  libasound2 libatk-bridge2.0-0 libatk1.0-0 libatspi2.0-0 libcups2 libnspr4 \
  libnss3 libxcomposite1 libxdamage1 libxkbcommon0 libxrandr2 libxtst6 xdg-utils \
  && dpkg -i package1.deb \
  package2.deb \
  package3.deb \
  && rm ./package* \
  && rm -rf /var/lib/apt/lists/*

RUN more stuff
```
A situation like this can easy gets unreadable and harder to change.

But with possible squash instructions, it could look like this:
```
SQUASH start

RUN wget -qO- "https://mystorage.com/some_debs_files.tar.xz" | tar xJf - -C .

RUN apt-get -qq update && apt-get -qq install -y \
  libasound2 libatk-bridge2.0-0 libatk1.0-0 libatspi2.0-0 libcups2 libnspr4 \
  libnss3 libxcomposite1 libxdamage1 libxkbcommon0 libxrandr2 libxtst6 xdg-utils

RUN dpkg -i package1.deb \
  package2.deb \
  package3.deb

RUN m ./package* \
  && rm -rf /var/lib/apt/lists/*

SQUASH end

RUN more stuff
```
This is easier to read, debug, and modify. The `SQUASH end` would only be executed if it's reached.

Now I'm guessing that the one thing that could cause problems here is the fact that `SQUASH start` and `SQUASH end` wouldn't act as layers per se.

Possible behaviors:
- `SQUASH start` with no `SQUASH end`: squash all layers after `SQUASH start`
- `SQUASH end` with no `SQUASH start`: error
- two `SQUASH start` without any `SQUASH end` between them: error
- `SQUASH start` without a `SQUASH end` wouldn't impact other images using this one as a base
- If the `SQUASH end` is not reached because some instructions failed, the work after the `SQUASH start` is cached for faster rebuilds while debugging

**Comment by benja-M-1** [comment]
Hi @jackcaron, I believe this should be posted at https://github.com/moby/buildkit rather than here. Would you mind creating an issue over there?
Ideas here must be related to extensions :) 
Thanks!

**Reply by jackcaron** [reply]
Thanks for the info.

Here's the new discussion if you want to follow it: https://github.com/moby/buildkit/discussions/3376
